CodeGeneration/Sources/generate-swiftsyntax/templates/swiftsyntax/SyntaxNodeFile.swift

```diff
@@ -0,0 +1,432 @@
+      for choice in child.nodeChoices {
+        StmtSyntax("""
+          if let node = node.as(\(raw: choice.typeName).self) {
+            self = .\(raw: choice.swiftName)(node)
+            return
+          }
+          """)
+      }
+
+      StmtSyntax("return nil")
+    }
+
+    try! VariableDeclSyntax("public static var structure: SyntaxNodeStructure") {
+      let choices = ArrayExprSyntax {
+        for choice in child.nodeChoices {
+          ArrayElementSyntax(
+            expression: ExprSyntax(".node(\(raw: choice.typeName).self)"))
+        }
+      }
+
+      StmtSyntax("return .choices(\(choices))")
+    }
+  }
+}
+
+private func generateInitializerDecl(children: [Child], optionalBaseAsMissing: Bool, currentIndentation: String = "") -> DeclSyntax {
+    if children.isEmpty {
+        return "public init()"
+    }
+
+    var genericParamNames: [String: Int] = [:]
+    var genericParams: [String] = []
+    var params = ["leadingTrivia: Trivia? = nil"]
+
+    for child in children {
+        var paramType: String
+        if !child.nodeChoices.isEmpty {
+            paramType = child.name
+        } else if child.hasBaseType {
+            if optionalBaseAsMissing {
+                paramType = "Missing\(child.typeName)"
+            } else {
+
+                // If we have a base type, make the initializer generic over its
+                // protocol instead.
+                let index = child.swiftName.index(child.swiftName.startIndex, offsetBy: 1)
+                paramType = child.swiftName[..<index].uppercased()
+
+                let paramCount = (genericParamNames[paramType] ?? 0) + 1
+                genericParamNames[paramType] = paramCount
+
+                if paramCount > 1 {
+                    paramType += "\(paramCount)"
+                }
+                genericParams.append("\(paramType): \(child.typeName)Protocol")
+            }
+        } else {
+            paramType = child.typeName
+        }
+
+        if child.isOptional {
+            paramType += "?"
+        }
+
+        let label = child.isUnexpectedNodes ? "_ " : ""
+
+        params.append("\(label)\(child.swiftName): \(paramType)\(generateDefaultInitialization(for: child))")
+    }
+
+    params.append("trailingTrivia: Trivia? = nil")
+
+    let paramIndentation = currentIndentation + "  "
+    let separator = ",\n" + paramIndentation
+    var generics = ""
+
+    if !genericParams.isEmpty {
+        generics = "<\(genericParams.joined(separator: ", "))>"
+    }
+
+    return """
+      public init\(raw: generics)(
+      \(raw: paramIndentation)\(raw: params.joined(separator: separator))
+      \(raw: currentIndentation))
+      """
+}
+
+/// If the type has a default value, return an expression of the form
+/// ` = default_value` that can be used as the default value to for a
+/// function parameter. Otherwise, return an empty string.
+private func generateDefaultInitialization(for child: Child) -> String {
```

Comment:
What do you think about making this an extension on `Child`?

```swift
fileprivate extension Child {
  var defaultInitialization: String { … }
}
```

Reply:
Should we also make the `generateInitializerDecl` above as an extension?
Then we don't need to pass the children list.

Reply:
Yes, I think that would make sense.